Switch to Barclay command line parser(s)

cmnbroad opened the pull request:
### Description

Changes to use the Barclay command line parser(s), as discussed [here](https://github.com/broadinstitute/picard/issues/786).

There are 5 separate commits to make this somewhat easier to review:

1. Changes to Barclay package/class names, annotations, attributes, and imports only.
2. Switch to the Barclay version of the existing parser. This mostly consists of removal of the existing command line parser, and changes to CommandLineProgram to use the Barclay version of the Picard parser.
3. Use @ArgumentCollection, remove overridable. With this commit, the tests pass using the Barclay version of the Picard parser.
4. Add ability to toggle between the two Barclay parsers. This includes property-based opt-in CLP toggling (default uses the existing parser); Gradle changes to generate a second artifact that uses the Barclay parser; minor changes to tools to get the tests to pass with both parsers, and a few changes required for tools that have no tests.
5. Gradle changes for the Barclay documentation generation process. These changes are build-only, plus the addition of the @DocumentedFeature annotation (required to make a tool a doc target) to use to verify this process.

There is one tool (DownsampleSam, and three more in picard-private) that use “R” as a command line argument for something other than reference, which is problematic for the @ArgumentCollection pattern. I’ve temporarily changed these to use a different short name. If thats  not workable, its possible to keep them as is, though it will require a little bit of ugly code and a typecast.

(There is also a separate branch for [picard-private](https://github.com/broadinstitute/picard-private/tree/cn_barclay) that has corresponding changes, subject to completion of this PR).

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [X] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [X] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

tfenne commented on src/main/java/picard/analysis/CollectWgsMetrics.java:
```diff
@@ -84,57 +86,74 @@
+     * @return An interval argument collection to be used for this tool. Subclasses can override this
+     * to provide an argument collection with alternative arguments or argument annotations.
+     */
+    protected IntervalArgumentCollection getIntervalArgumentCollection() {
```
I wonder if methods like these should be called `build...` or `make...` to make it clear that they're constructing a collection object to be used, not accessing the one that is populated by Barclay?

cmnbroad replied:
Yeah, "make" seems like a better choice. Done.

tfenne commented on src/main/java/picard/analysis/CollectWgsMetrics.java:
```diff
@@ -84,57 +86,74 @@
+        return new CollectWGSMetricsIntervalArgumentCollection();
+    }
+
+    public static class CollectWGSMetricsIntervalArgumentCollection implements IntervalArgumentCollection {
```
Could this be `CollectWgsMetricsIntervalArgumentCollection` to be consistent with the enclosing class please?

cmnbroad replied:
Done.

tfenne commented on src/main/java/picard/analysis/CollectWgsMetrics.java:
```diff
@@ -84,57 +86,74 @@
-            "all bases of reads that overlap these intervals will be considered, even if some of those bases extend beyond the boundaries of " +
-            "the interval. The ideal use case for this argument is to use it to restrict the calculation to a subset of (whole) contigs.",
-            optional = true, overridable = true)
     public File INTERVALS = null;
```
Same here .. I think this needs to be protected.  If anyone uses this class programmatically from another tool, and sets `INTERVALS` directly, it'll get clobbered by the value from `intervalArugmentCollection` via the assignment on line 450.  Better to get a compile break than a runtime break I think.

cmnbroad replied:
Made protected.

tfenne commented on src/main/java/picard/analysis/directed/CollectTargetedMetrics.java:
```diff
@@ -60,42 +60,47 @@ protected abstract COLLECTOR makeCollector(final Set<MetricAccumulationLevel> ac
     public int NEAR_DISTANCE = TargetedPcrMetricsCollector.NEAR_PROBE_DISTANCE_DEFAULT;
 
-    @Option(shortName = MINIMUM_MAPPING_QUALITY_SHORT_NAME, doc = "Minimum mapping quality for a read to contribute coverage.", overridable = true)
+    //TODO: MINIMUM_MAPPING_QUALITY and MINIMUM_BASE_QUALITY are both annotated as required,
```
I think the intent is that they are "required" but that the requirement is met by having a default value.  I'm not sure what the right way to represent that is in Barclay, but IIRC Picard's parser treats required-ness as "required to have a value after parsing", whether that value came from a default, or came from a command line arg.  I wouldn't change these particular ones to nullable values.

cmnbroad replied:
Barclay retains this, um, feature, and works the same way. It causes some [problems](https://github.com/broadinstitute/barclay/issues/23) though so I'd like to change it at some point.

Anyway, for now Ill make no change other than to remove my comment.

tfenne commented on src/main/java/picard/analysis/replicates/CollectIndependentReplicateMetrics.java:
```diff
@@ -102,50 +102,50 @@
         programGroup = Alpha.class
 )
 public class CollectIndependentReplicateMetrics extends CommandLineProgram {
     private static final int DOUBLETON_SIZE = 2, TRIPLETON_SIZE = 3;
```
I saw this line and said to myself, "this has to be a @yfarjoun class", and sure enough... 😉

tfenne commented on src/main/java/picard/cmdline/CommandLineProgram.java:
```diff
@@ -65,45 +80,57 @@
-    }
-
-
+ *       new MyConcreteClass().instanceMain(argv);
```
Two things: the documentation above is now out of date (it references the old annotations).  Also it's been years since Picard required `main()` methods in it's classes, I'm assuming Barclay doesn't either?  I.e. this has been inaccurate for a long time, not just made so by this commit.

cmnbroad replied:
Fixed.

tfenne commented on src/main/java/picard/cmdline/CommandLineProgram.java:
```diff
@@ -65,45 +80,57 @@
+    private static String PROPERTY_CONVERT_LEGACY_COMMAND_LINE = "picard.convertCommandLine";
+    private static Boolean useLegacyParser;
+
+    @Argument(common=true, optional=true)
```
Does this get `doc` some other way or should it have something here?

cmnbroad replied:
Updated with doc.

tfenne commented on src/main/java/picard/cmdline/CommandLineProgram.java:
```diff
@@ -65,45 +80,57 @@
     public ValidationStringency VALIDATION_STRINGENCY = ValidationStringency.DEFAULT_STRINGENCY;
 
-    @Option(doc = "Compression level for all compressed files created (e.g. BAM and GELI).", common=true)
+    @Argument(doc = "Compression level for all compressed files created (e.g. BAM and GELI).", common=true)
```
Another archaic doc that could be fix.  GELI is long gone.  Could you fix this please to say `e.g. BAM and .gz files`?

cmnbroad replied:
Done.

tfenne commented on src/main/java/picard/cmdline/CommandLineProgram.java:
```diff
@@ -254,18 +279,30 @@ public int instanceMain(final String[] argv) {
+            ret = commandLineParser.parseArguments(System.err, argv);
+        } catch (CommandLineException e) {
+            // Barclay command line parser throws on parsing/argument errors
+            System.err.println(e.getMessage());
```
Yes please to reversing these two lines!

cmnbroad replied:
Done.

tfenne commented on src/main/java/picard/cmdline/CommandLineProgram.java:
```diff
@@ -254,18 +279,30 @@ public int instanceMain(final String[] argv) {
                 System.err.println(msg);
             }
-            commandLineParser.usage(System.err, false);
+            System.err.print(commandLineParser.usage(false, false));
```
Ditto here.

cmnbroad replied:
Done.

tfenne commented on src/main/java/picard/cmdline/CommandLineProgram.java:
```diff
@@ -311,20 +383,38 @@ public void setDefaultHeaders(final List<Header> headers) {
+    }
+
+    /**
+     * Determine if a class has wbe documentation based on its package name
```
wbe -> web

cmnbroad replied:
Fixed.

tfenne commented on src/main/java/picard/cmdline/PicardCommandLine.java:
```diff
@@ -241,9 +243,9 @@ private static void printUsage(final Set<Class<?>> classes, final String command
                 if (!commandListOnly) {
                     if (clazz.getSimpleName().length() >= 45) {
-                        builder.append(String.format("%s    %s    %s%s%s\n", KGRN, clazz.getSimpleName(), KCYN, property.usageShort(), KNRM));
+                        builder.append(String.format("%s    %s    %s%s%s\n", KGRN, clazz.getSimpleName(), KCYN, property.summary(), KNRM));
```
Shouldn't this also be `oneLineSummary`?

cmnbroad replied:
Hm, yes it should....Done.

magicDGS commented on build.gradle:
```diff
@@ -81,11 +89,46 @@ jar {
 }
 
+// Generate Picard Online Doc
+task picardDoc(type: Javadoc, dependsOn: ['cleanPicardDoc', classes]) {
```
Will be nice to have some kind of plugin for generating the documentation with Barclay without repeating code here and in GATK (and my own software). Is there any way to do it easier?

lbergelson replied:
@magicDGS A gradle plugin for barclay seems like a good idea.  I've never written one, but I imagine it's probably not that difficult.  I opened a ticket to discuss it #819

lbergelson replied:
Ack, I put that in the wrong place, see https://github.com/broadinstitute/barclay/issues/56 instead.

magicDGS commented on src/main/java/picard/util/help/HelpConstants.java:
```diff
@@ -0,0 +1,69 @@
+
+    private static Map<String, String> groupToSuperCategory;
+
+    private static Map<String, String> getSuperCategoryMap() {
```
It will be nice to have this method and the supercategory values accessible to reuse this map. For example, if I create a new program group for a tool I can add it to the tools supercategory and like that reuse this class to also include the others without copying the code in this method and re-define the supercat constants....

I would like to have this also in GATK to re-use super-categories and read groups...

cmnbroad replied:
@magicDGS We can certainly widen the visibility (here and in GATK), assuming you mean that you want to be able to add entries to this from your custom doclet ?

magicDGS replied:
Exactly. Either add entries from my custom doclet or use as default and override if necessary some of the supercategories (by creating a new one, not directly here, to be safer).

cmnbroad commented:
@yfarjoun Can you or review or appoint someone ? Thx.

yfarjoun commented on src/main/java/picard/util/help/PicardHelpDoclet.java:
```diff
@@ -0,0 +1,110 @@
+ */
+public class PicardHelpDoclet extends HelpDoclet {
+
+    private final static String PICArD_FREEMARKER_INDEX_TEMPLATE_NAME = "generic.index.template.html";
```
why the lowercase `r` in PICArD?

cmnbroad replied:
Typo. Fixed.

jacarey commented on src/main/java/picard/cmdline/CommandLineProgram.java:
```diff
@@ -254,18 +279,30 @@ public int instanceMain(final String[] argv) {
-        final boolean ret = commandLineParser.parseOptions(System.err, argv);
+        commandLineParser = getCommandLineParser();
+
+        boolean ret = false;
```
redundant set

cmnbroad replied:
Fixed.

jacarey commented on src/main/java/picard/cmdline/CommandLineProgram.java:
```diff
@@ -282,13 +319,48 @@ protected boolean parseArgs(final String[] argv) {
+                        new LegacyCommandLineArgumentParser(this) :
+                        new CommandLineArgumentParser(this,
+                            Collections.EMPTY_LIST,
+                            new HashSet<>(Arrays.asList(CommandLineParserOptions.APPEND_TO_COLLECTIONS)));
```
Collections.singletonList

pshapiro4broad replied:
Better yet replace the expression with `Collections.singleton(CommandLineParserOptions.APPEND_TO_COLLECTIONS))` or
`EnumSet.of(CommandLineParserOptions.APPEND_TO_COLLECTIONS)`. Both assume that `CommandLineArgumentParser` doesn't side effect its arguments, which it shouldn't.

cmnbroad replied:
Done.

jacarey commented on src/main/java/picard/cmdline/CommandLineProgram.java:
```diff
@@ -282,13 +319,48 @@ protected boolean parseArgs(final String[] argv) {
+     * @return true if the legacy parser should be used
+     */
+    public static boolean useLegacyParser(final Class<?> clazz) {
+        String legacyPropertyValue = null;
```
can be scoped inside the if and doesn't need initialization

cmnbroad replied:
Done.

jacarey commented on src/main/java/picard/util/help/HelpConstants.java:
```diff
@@ -0,0 +1,69 @@
+
+public final class HelpConstants {
+
+    private HelpConstants() {};
```
extra semicolon

cmnbroad replied:
Done.

jacarey commented on src/test/java/picard/cmdline/PicardCommandLineTest.java:
```diff
@@ -13,4 +23,40 @@ public void TestPicardPublic() { // this tests fails if any CLP in picard is mis
+            classFinder.find(pkg, CommandLineProgram.class);
+        }
+
+        final Map<String, Class<?>> simpleNameToClass = new HashMap<String, Class<?>>();
```
unused variable

cmnbroad replied:
Done.

yfarjoun commented:
@cmnbroad this seems to be waiting for you to respond and rebase...

cmnbroad commented:
Yes, I expect to get back to this sometime this week.

cmnbroad commented on src/main/java/picard/analysis/CollectRrbsMetrics.java:
```diff
@@ -127,6 +123,20 @@
+
+    public static class CollectRrbsMetricsReferenceArgumentCollection implements ReferenceArgumentCollection {
+        @Argument(doc = "The reference sequence fasta file", shortName = StandardOptionDefinitions.REFERENCE_SHORT_NAME)
+        public File REFERENCE;
```
Barclay doesn't directly support either of these now, but I think it could be added if we wanted to go down this path since the underlying jopt library appears to support both abbreviations and multiple names.

cmnbroad commented on src/main/java/picard/cmdline/CommandLineProgram.java:
```diff
@@ -65,45 +80,57 @@
+    @ArgumentCollection
+    public ReferenceArgumentCollection referenceSequence = getReferenceArgumentCollection();
+
     public File REFERENCE_SEQUENCE = Defaults.REFERENCE_FASTA;
```
Yes, I should have been more explicit about why I kept that. And I see your point (articulated in a comment below), about these possibly being clobbered, but CollectMultipleMetrics sets REFERENCE_SEQUENCE directly for all of the CLPs that it uses. I've reduced the visibility to protected, and added a setReferenceSequence method to SinglePassSam so that CollectMultipleMetrics can populate this, which reduces the scope of the exposure. Let me know if you have another suggestion.

On a somewhat unrelated note: in GATK4, we had previously refactored (some) of the metrics collectors that originally came from Picard in a way thats similar to what @nh13 did [here](https://github.com/broadinstitute/picard/pull/840), only more so. For each type of metric, we factored out:

1) an accumulator class for holding the results (which is also a reducer, where applicable)
2) an argument collection class for the controlling arguments
3) the collector itself, which accepts a read and populates the accumulator

In addition to codifying the ability to set the arguments for a collector independently of it's container, it entirely separates the metrics computation from the source of the reads (which in GATK's case could be a CLP, a multiple collector CLP, a Spark CLP, or a Spark RDD in a pipeline). The CLPs are just thin shells around these components. Its obsolete now since once this Barclay PR goes through we'll use the Picard version of the collectors, but I would lobby that we adopt a similar approach similar in Picard.

cmnbroad commented:
@tfenne @jacarey @pshapiro4broad @yfarjoun Sorry for the delay on follow up. I've added one commit with changes based on all of the review comments. Once everyone is good with those, I'd like to squash all of these commits down to a single commit before I rebase (this is now about a month behind master), since its painful to keep the changes in the individual commits in sync through a rebase. Once I've rebased I'll make a separate commit with any new changes required at that point. Finally, I'll update my corresponding picard-private branch and make the PR for that.

nh13 commented:
@cmnbroad this is great! When you have rebased, I'd like to build a copy and make sure that it "works" in my hands and some of my pipelines.  Thank-you for all this effort.

cmnbroad commented:
@ktibbett @yfarjoun Also, as I mentioned in the original PR comments, in this PR I had to change the "R" argument for DownsampleSam (which uses "R" for the random seed) since it collides with the REFERENCE short name). Please let me know if you guys *really* can't live with that, in which case I'll have to add a little bit of ugly coding/typecasting to preserve the existing command line arg name.

vdauwera commented:
@cmnbroad Re: the DownsampleSam  random seed thing, what happens if the user doesn't update their command accordingly? I think the desired behavior should be to fail immediately with an informative error message ("yo, update your command, buddy -- R means reference now; the new arg name for randmo seed is [blah]"). If it fails late or screws up silently then we'd want to put in some logic to catch the case.

cmnbroad commented:
@vdauwera Good question. I just tried it, and it fails early, but badly (with an exception/stacktrace), with either parser. If we keep it, I can add in some code that checks to see if the reference you specified looks more like a random seed than a filename, and fail a bit more gracefully.

cmnbroad commented:
@nh13 A test drive before we merge this would be great - I'll ping you when its rebased.

cmnbroad commented:
@nh13 This is rebased and up-to-date on current Picard as of today if you want to try it out.

cmnbroad commented:
@vdauwera Also, I removed the DownsampleSam issue, so this is compatible now with the existing command line.

pshapiro4broad commented:
Looks ok to me 👍

lbergelson commented on build.gradle:
```diff
@@ -94,6 +165,46 @@ task currentJar(type: Copy){
+    systemProperty 'picard.useLegacyParser', 'false'
+}
+
+test {
```
Could you add a `legacyTest(type: Test)` target and make test depend on both that and `barclayTest` so that it's possible to run just one or the other.

cmnbroad replied:
Sure.

lbergelson commented on build.gradle:
```diff
@@ -94,6 +165,46 @@ task currentJar(type: Copy){
+// file that tells Picard to use the Barclay command line parser instead of the Picard
+// command line parser.
+//
+task barclayJar(type: com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar)  {
```
Shouldn't this be `barclayShadowJar`?

cmnbroad replied:
Yes, makes sense.

lbergelson commented on build.gradle:
```diff
@@ -257,6 +276,9 @@ task sourcesJar(type: Jar) {
 artifacts {
+    // Make sure the barclayJar is listed first to prevent it from being chosen as the one to be
+    // installed under the snapshot name in the local maven repository when doing "./gradlew install".
+    archives barclayJar
```
I think we want to upload a non-shadow jar as the barclay jar we archive so that gatk can depend on it.  We probably need to add another jar task to make that.

lbergelson commented on build.gradle:
```diff
@@ -322,25 +344,3 @@ uploadArchives {
 }
-
-//update static web docs
-task copyJavadoc(dependsOn: 'javadoc', type: Copy) {
```
Are we ok with no longer updating the picard website?

cmnbroad replied:
@vdauwera We've discussed this in the past, but just want to confirm: once we commit this branch, the only doc generation capability we'll have for Picard will be the equivalent of gatkDoc, which will generate Barclay-style syntax, and will require a the addition of @DocumentedFeature annotations).

vdauwera replied:
@cmnbroad Sorry for the delayed response, I was out for a while then catching up. 

I can do the addition of @documentedfeature annotations. Should that be part of this pull request or one queued up to follow right away afterward? 

We do need to set up the release system to build and publish the gatkDocs on the http://broadinstitute.github.io/picard/ website, as I don't think anyone is ready yet to have Picard docs on the GATK website only.  

Github now provides a simpler option for publishing a doc website; based on [this post](https://github.com/blog/2228-simpler-github-pages-publishing) it looks like we just need to switch a repo setting and we'll be able to simply copy the docs in a `/docs` directory. I'll edit the template to work with the picard theme. 

Am I missing anything?

lbergelson commented on build.gradle:
```diff
@@ -78,12 +107,54 @@ jar {
                 'Implementation-Vendor': 'Broad Institute',
                 'Implementation-Version': version
     }
+    from('src/main/resources/properties.templates') {
```
This repeated copy block is so ugly.  I'm trying to come up with a way around it but I'm not sure how to do it in a way that isn't worse.

cmnbroad replied:
Yeah, I'm sure there is some way. Let me know if you have something.

cmnbroad commented on build.gradle:
```diff
@@ -78,12 +107,54 @@ jar {
+}
+
+javadoc {
+    // This is a hack to disable the java 8 default javadoc lint until we fix the html formatting
```
Done.

cmnbroad commented on build.gradle:
```diff
@@ -78,12 +107,54 @@ jar {
+javadoc {
+    // This is a hack to disable the java 8 default javadoc lint until we fix the html formatting
+    // We only want to do this for the javadoc task, not gatkDoc
+    options.addStringOption('Xdoclint:none', '-quiet')
```
Without it, the javadoc task causes the build to fail due to javadoc issues.

cmnbroad commented on build.gradle:
```diff
@@ -94,6 +165,46 @@ task currentJar(type: Copy){
+    configurations = [project.configurations.runtime]
+    from project.sourceSets.main.output
+    manifest {
+        inheritFrom project.tasks.jar.manifest
```
Yeah - I think I could also move the manifest 'attributes' into that as well.

cmnbroad commented on build.gradle:
```diff
@@ -94,6 +165,46 @@ task currentJar(type: Copy){
+            fileName.replace(barclaySourcePropertyFile, picardTargetPropertyFile)
+        }
+    }
+    exclude('META-INF/INDEX.LIST', 'META-INF/*.SF', 'META-INF/*.DSA', 'META-INF/*.RSA')
```
Good catch - I don't recall what made me add this, but I don't think we need it here. Removed.

cmnbroad commented:
@lbergelson Ping - I'm ready to rebase this on current code once you review the gradle changes I made based on your comments.

cmnbroad commented:
@ktibbett @jacarey @vdauwera @nh13 @tfenne @pshapiro4broad I think all the stars are finally aligned, and I’m planning to squash and rebase this one more time, and merge, on Friday. Please let me know if there any concerns with that plan.

Note (@ktibbett @jacarey) that once this branch is merged, picard-private can’t be upgraded to a new picard build again until the corresponding picard-private [PR](https://github.com/broadinstitute/picard-private/pull/987) is also merged. In order to minimize the window of time where the two repos are out of sync, I’d like to get that PR reviewed as soon as possible. Can you assign someone to do that ? It passes tests locally, but won’t actually build on Jenkins until the this PR is merged and the Picard snapshot is available via artifactory.

vdauwera commented:
We need a bit more wiring to migrate the docs publishing, see my comment above.
